### PR TITLE
[FLINK-34689][MySQL][Feature] check  binlog_row_value_options

### PR DESCRIPTION
**Pull request by SML0127**
Implemented functino to check `binlog_row_value_optoins` within MySqlValidator.

When `binlog_row_value_optoins` is set to `PARTIAL_JSON`,  
the update operator remains as `Update_rows_partial`.
pasted_image_at_2024-03-07T03_21_08 548Z

Flink CDC does not parse this event because `Update_row_partial` binlog event is mapped to `PARTIAL_UPDATE_ROWS_EVENT` and Flink CDC do not handle that event type. Therefore, some update query-related binary logs are lost.

So, we have to check `binlog_row_value_optoins` before starting.

### Implemented checkBinlogRowValueOptions
```java
private final MySqlSourceConfig sourceConfig;
    public void validate() {
        checkVersion(connection);
        checkBinlogFormat(connection);
        checkBinlogRowImage(connection);
        checkBinlogRowValueOptions(connection);
        checkTimeZone(connection);
    } catch (SQLException ex) {
        throw new TableException(
    }
}

/** Check whether the binlog row value options is empty. */
private void checkBinlogRowValueOptions(JdbcConnection connection) throws SQLException {
    String rowValueOptions =
           connection
                    .queryAndMap(
                            "SHOW GLOBAL VARIABLES LIKE 'binlog_row_value_options'",
                            rs -> rs.next() ? rs.getString(2) : "")
                    .trim()
                    .toUpperCase();
    // This setting was introduced in MySQL 8.0+ with default of empty string ''
    // For older versions, assume empty string ''
    if (!rowValueOptions.equals(BINLOG_ROW_VALUE_OPTIONS)) {
        throw new ValidationException(
                String.format(
                       "The MySQL server is configured with binlog_row_value_options %s rather than %s, which is "
                                + "required for this connector to work properly. Change the MySQL configuration to use a "
                                + "binlog_row_image='' and restart the connector.",
                        rowValueOptions, BINLOG_ROW_VALUE_OPTIONS));
    }
}
```

### CODE REVIEW

**Comment by SML0127** [comment]
jira issue: https://issues.apache.org/jira/browse/FLINK-34689

**Comment by SML0127** [comment]
@leonardBang , @ruanhang1993 CC

**Comment by leonardBang** [comment]
Thanks @SML0127 for the contribution, @ruanhang1993 Would you like to help review this PR?

**Comment by SML0127** [comment]
@ruanhang1993 PTAL

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 1]
```diff
@@ -159,6 +161,27 @@ private void checkBinlogRowImage(JdbcConnection connection) throws SQLException
...
+                        .toUpperCase();
+        // This setting was introduced in MySQL 8.0+ with default of empty string ''
+        // For older versions, assume empty string ''
+        if (!rowValueOptions.equals(BINLOG_ROW_VALUE_OPTIONS)) {
+            throw new ValidationException(
+                    String.format(
+                            "The MySQL server is configured with binlog_row_value_options %s rather than %s, which is "
+                                    + "required for this connector to work properly. Change the MySQL configuration to use a "
+                                    + "binlog_row_image='' and restart the connector.",
+                            rowValueOptions, BINLOG_ROW_VALUE_OPTIONS));
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:181]: I think this error message should be changed as follows.

`The MySQL server is configured with binlog_row_value_options=%s, which is possible to cause losing some binlog events for the mysql cdc connector. Please change the MySQL configuration to use a  binlog_row_image='' and rerun the job. See more details at https://dev.mysql.com/doc/refman/8.0/en/replication-features-json.html.`

**Review by ruanhang1993 — COMMENTED** [review]
@SML0127 Hi. Sorry for the late review. I left a little comment. Thanks~

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java (hunk 1 above)
Comment (ruanhang1993) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:181]: If I do not understand wrong, we do not need to change the binlog_row_image setting. Am I right ?

Comment (SML0127) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:181]: > If I do not understand wrong, we do not need to change the binlog_row_image setting. Am I right ?

@ruanhang1993 
I updated log message. 

Yes. we do not need to change both settings `binlog_row_image` and `binlog_row_value_options`.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 2]
```diff
@@ -50,6 +50,7 @@ public class MySqlValidator implements Validator {
 
     private static final String BINLOG_FORMAT_ROW = "ROW";
     private static final String BINLOG_FORMAT_IMAGE_FULL = "FULL";
+    private static final String BINLOG_ROW_VALUE_OPTIONS = "";
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:53]:
```suggestion
    private static final String DEFAULT_BINLOG_ROW_VALUE_OPTIONS = "";
```

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 3]
```diff
@@ -159,6 +161,27 @@ private void checkBinlogRowImage(JdbcConnection connection) throws SQLException
         }
     }
 
+    /** Check whether the binlog row value options is empty. */
+    private void checkBinlogRowValueOptions(JdbcConnection connection) throws SQLException {
+        String rowValueOptions =
+                connection
+                        .queryAndMap(
+                                "SHOW GLOBAL VARIABLES LIKE 'binlog_row_value_options'",
+                                rs -> rs.next() ? rs.getString(2) : "")
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:170]:
```suggestion
                                rs -> rs.next() ? rs.getString(2) : DEFAULT_BINLOG_ROW_VALUE_OPTIONS)
```

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 4]
```diff
@@ -159,6 +161,27 @@ private void checkBinlogRowImage(JdbcConnection connection) throws SQLException
...
+                        .trim()
+                        .toUpperCase();
+        // This setting was introduced in MySQL 8.0+ with default of empty string ''
+        // For older versions, assume empty string ''
+        if (!rowValueOptions.equals(BINLOG_ROW_VALUE_OPTIONS)) {
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:175]:
```suggestion
        if (!BINLOG_ROW_VALUE_OPTIONS.equals(rowValueOptions)) {
```

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 5]
```diff
@@ -159,6 +161,27 @@ private void checkBinlogRowImage(JdbcConnection connection) throws SQLException
...
+            throw new ValidationException(
+                    String.format(
+                            "The MySQL server is configured with binlog_row_value_options=%s, which is possible to cause losing some binlog events"
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:178]:
```suggestion
                            "The MySQL server is configured with binlog_row_value_options=%s, which is possible to cause losing some binlog events "
```

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 6]
```diff
@@ -159,6 +161,27 @@ private void checkBinlogRowImage(JdbcConnection connection) throws SQLException
...
+                                    + "for the mysql cdc connector. Please remove the binlog_row_value_options setting in the MySQL server and rerun the job."
+                                    + "See more details at https://dev.mysql.com/doc/refman/8.0/en/replication-features-json.html.",
+                            rowValueOptions, BINLOG_ROW_VALUE_OPTIONS));
```
Comment (ruanhang1993) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:181]:
```suggestion
                            rowValueOptions));
```

**Review by ruanhang1993 — COMMENTED** [review]
Hi, @SML0127.
 I left some comments. Please take a look, thanks~

**Comment by SML0127** [comment]
> Hi, @SML0127. I left some comments. Please take a look, thanks~

@ruanhang1993 
Thank you for your comments!
I have applied all of your comments to the code.

File: flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java [hunk 7]
```diff
@@ -50,6 +50,7 @@ public class MySqlValidator implements Validator {
 
     private static final String BINLOG_FORMAT_ROW = "ROW";
     private static final String BINLOG_FORMAT_IMAGE_FULL = "FULL";
```
Comment (SML0127) [inline flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:52]: @ruanhang1993 
May I change the variable names `BINLOG_FORMAT_ROW`, `BINLOG_FORMAT_IMAGE_FULL` to `DEFAULT_BINLOG_FORMAT`, `DEFAULT_BINLOG_ROW_FORMAT`?

```suggestion
    private static final String DEFAULT_BINLOG_FORMAT = "ROW";
    private static final String DEFAULT_BINLOG_ROW_FORMAT = "FULL";
```


In mysql 8, default values for the `binlog_format` and `binlog_row_image` options are `ROW` and `full`, respectively
- https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_binlog_format
- https://dev.mysql.com/doc/refman/8.0/en/replication-options-binary-log.html#sysvar_binlog_row_image

Comment (ruanhang1993) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:52]: Hi, @SML0127 .
I think we could not change these fields' names. These settings are checked by the `ROW` or `FULL` values.

Comment (SML0127) [reply flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/org/apache/flink/cdc/connectors/mysql/MySqlValidator.java:52]: Thank you for your explanation, @ruanhang1993 
I undertand your responses. Do you have any other suggestions?

**Review by ruanhang1993 — APPROVED** [review]
LGTM.

**Review by PatrickRen — APPROVED** [review]
@SML0127 Thanks for the contribution! LGTM

**Pull request merged**